File: 002_even_fibonacci_numbers/even_fibonacci_numbers.rb
def sum_of_evens(limit)

	previous, current, start_evens_sum = [0, 1, 0]
	sequence(previous, current, start_evens_sum, limit)

end

def sequence(previous, current, evens_sum, limit)

	return evens_sum if current >= limit
	sum = previous + current
	even?(sum) ? new_sum = evens_sum + sum : new_sum = evens_sum
	sequence(current, sum, new_sum, limit)

end

def even?(number)
	number % 2 == 0 ? true : false
end

# Tracks time
start = Time.now
puts sum_of_evens(4000000)
puts "Execution time: #{Time.now - start}"

# Answer: 4613732
# Time Elasped: 0.000048

File: 002_even_fibonacci_numbers/even_fibonacci_numbers.ex
defmodule EvenFibonacciNumbers do

  def evens_sum(limit), do: IO.puts "Answer: #{do_fib(0,1,0, limit)}"

  defp do_fib(_, current, result, limit) when current >= limit, do: result

  defp do_fib(prev, current, result, limit) when current < limit do
    sum = prev + current
    if rem(sum, 2) == 0, do: result = sum + result
    do_fib(current, sum, result, limit)
  end

end

# Used to track time
defmodule Benchmark do

  def measure(function) do
    function
    |> :timer.tc
    |> elem(0)
    |> Kernel./(1_000_000)
  end

end

IO.puts "Execution time: #{Benchmark.measure(fn -> 
  EvenFibonacciNumbers.evens_sum(4_000_000) end)}"

# Answer: 4613732
# Execution time: 0.001753

